go-controller/pkg/ovn/healthcheck/egressip_healthcheck.go

```diff
@@ -0,0 +1,174 @@
+		if err := grpcServer.Serve(lis); err != nil && err != grpc.ErrServerStopped {
+			klog.Fatalf("Health checking serve failed: %v", err)
+		}
+		klog.Infof("Stopped Egress IP Health Server on %s:%d", ehs.nodeMgmtIP.String(), ehs.healthCheckPort)
+	}()
+
+	<-stopCh
+
+	klog.Info("Shutting down Egress IP Health Server")
+	grpcServer.Stop()
+	wg.Wait()
+	klog.Info("Shut down Egress IP Health Server")
+}
+
+// EgressIPHealthClient interface offers the functions needed for connecting to
+// the egress ip health check service
+type EgressIPHealthClient interface {
+	IsConnected() bool
+	Connect(dialCtx context.Context, mgmtIPs []net.IP, healthCheckPort int) bool
+	Disconnect()
+	Probe(dialCtx context.Context) bool
+}
+
+type egressIPHealthClient struct {
+	nodeName string
+	nodeAddr string
+	conn     *grpc.ClientConn
+	// the probeFailed state is used to mitigate situations when
+	// connection just went down. With that, we do not declare node
+	// unreachable unless connection could not be re-established
+	probeFailed bool
+}
+
+// NewEgressIPHealthClient allocates an Egress IP health client
+func NewEgressIPHealthClient(nodeName string) EgressIPHealthClient {
+	return &egressIPHealthClient{nodeName: nodeName}
+}
+
+func (ehc *egressIPHealthClient) IsConnected() bool {
+	return ehc.conn != nil
+}
+
+func (ehc *egressIPHealthClient) Connect(dialCtx context.Context, mgmtIPs []net.IP, healthCheckPort int) bool {
+	var conn *grpc.ClientConn
+	var nodeAddr string
+	var err error
+
+	for _, nodeMgmtIP := range mgmtIPs {
+		options := []grpc.DialOption{
+			grpc.WithBlock(),
+			// TODO (FF): replace with TLS credentials
+			grpc.WithTransportCredentials(insecure.NewCredentials()),
+		}
+		nodeAddr = fmt.Sprintf("%s:%d", nodeMgmtIP.String(), healthCheckPort)
+		conn, err = grpc.DialContext(dialCtx, nodeAddr, options...)
+		if err == nil && conn != nil {
+			break
+		}
+	}
+	if conn == nil {
+		klog.Warningf("Could not connect to %s (%s): %v", ehc.nodeName, nodeAddr, err)
+		return false
+	}
+
+	klog.Infof("Connected to %s (%s)", ehc.nodeName, nodeAddr)
+	ehc.nodeAddr = nodeAddr
+	ehc.conn = conn
+	return true
+}
+
+func (ehc *egressIPHealthClient) Disconnect() {
+	if ehc.conn != nil {
+		klog.Infof("Closing connection with %s (%s)", ehc.nodeName, ehc.nodeAddr)
+		ehc.conn.Close()
+		ehc.conn = nil
+	}
+}
+
+func (ehc *egressIPHealthClient) Probe(dialCtx context.Context) bool {
+
+	if ehc.conn == nil {
+		// should never happen
+		klog.Warningf("Unexpected probing before connecting %s", ehc.nodeName)
+		return false
+	}
+
+	response, err := NewHealthClient(ehc.conn).Check(dialCtx, &HealthCheckRequest{Service: serviceEgressIPNode})
+	if err != nil {
+		// check failed. What we will return here will depend on ehc.probe_failed. If this is the first failure,
+		// let's tolerate it to account for cases where session went down and we just need it re-established.
+		// Otherwise, declare it failed.
```

Comment:
Just a note for myself: Ok so this means that we tolerate 1 lost packet, but with the second probe that fails, take it down. We probe every 5 seconds (previous commit)

Reply:
That is right. In cases where we fail back to back (right after re-connecting) we will also declare failure right away too.
Hope that is not too confusing.